edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/LogsRequestHandler.cs

```diff
@@ -0,0 +1,99 @@
+        {
+            LogsRequest payload = payloadOption.Expect(() => new ArgumentException("Request payload not found"));
+            ILogsRequestToOptionsMapper requestToOptionsMapper = new LogsRequestToOptionsMapper(
+                this.runtimeInfoProvider,
+                payload.Encoding,
+                payload.ContentType,
+                LogOutputFraming.None,
+                Option.None<LogsOutputGroupingConfig>(),
+                false);
+
+            IList<(string id, ModuleLogOptions logOptions)> logOptionsList = await requestToOptionsMapper.MapToLogOptions(payload.Items, cancellationToken);
+            IEnumerable<Task<LogsResponse>> uploadLogsTasks = logOptionsList.Select(
+                async l =>
+                {
+                    Events.ReceivedLogOptions(l);
+                    ModuleLogOptions logOptions = l.logOptions.Filter.Tail
+                        .Filter(t => t < MaxTailValue)
+                        .Map(t => l.logOptions)
+                        .GetOrElse(
+                            () =>
+                            {
+                                var filter = new ModuleLogFilter(Option.Some(MaxTailValue), l.logOptions.Filter.Since, l.logOptions.Filter.LogLevel, l.logOptions.Filter.RegexString);
+                                return new ModuleLogOptions(l.logOptions.ContentEncoding, l.logOptions.ContentType, filter, l.logOptions.OutputFraming, l.logOptions.OutputGroupingConfig, l.logOptions.Follow);
+                            });
+
+                    byte[] moduleLogs = await this.logsProvider.GetLogs(l.id, logOptions, cancellationToken);
+
+                    Events.ReceivedModuleLogs(moduleLogs, l.id);
+                    return logOptions.ContentEncoding == LogsContentEncoding.Gzip
+                        ? new LogsResponse(l.id, moduleLogs)
+                        : new LogsResponse(l.id, moduleLogs.FromBytes());
+                });
+            IEnumerable<LogsResponse> response = await Task.WhenAll(uploadLogsTasks);
+            return Option.Some(response);
+        }
+
+        static class Events
+        {
+            const int IdStart = AgentEventIds.LogsRequestHandler;
+            static readonly ILogger Log = Logger.Factory.CreateLogger<LogsRequestHandler>();
+
+            enum EventIds
+            {
+                ReceivedModuleLogs = IdStart + 1,
+                ReceivedLogOptions
+            }
+
+            public static void ReceivedModuleLogs(byte[] moduleLogs, string id)
+            {
+                Log.LogInformation((int)EventIds.ReceivedModuleLogs, $"Received {moduleLogs.Length} bytes of logs for {id}");
+            }
+
+            public static void ReceivedLogOptions((string id, ModuleLogOptions logOptions) receivedLogOptions)
+            {
+                if (receivedLogOptions.logOptions.Filter.Tail.HasValue)
```

Comment:
Nit: You have this pattern:
```
if (Tail.HasValue) {
  Tail.ForEach(t => ...);
} else {
  ...
}
```

Would `Tail.Match(t => ..., () => ...)` be better?

Reply:
Option.Match expects a result, which is why I used this pattern instead (as the return type here is void). Maybe we can add a new construct to Option that doesn't expect a type to be returned.

Reply:
Ah, ok. This is fine.